src/Microsoft.ML.DataView/TypeManager.cs

```diff
@@ -0,0 +1,189 @@
+        /// </summary>
+        private static Dictionary<int, Type> _idToTypeMap = new Dictionary<int, Type>();
+
+        /// <summary>
+        /// Mapping from ID to a <see cref="DataViewType"/> instance. The ID is the ID of <see cref="DataViewType"/> instance in ML.NET's type system.
+        /// </summary>
+        private static Dictionary<int, DataViewType> _idToDataViewTypeMap = new Dictionary<int, DataViewType>();
+
+        /// <summary>
+        /// Mapping from hashing ID of a <see cref="Type"/> and its <see cref="Attribute"/>s to hashing ID of a <see cref="DataViewType"/>.
+        /// </summary>
+        private static Dictionary<int, int> _typeIdToDataViewTypeIdMap = new Dictionary<int, int>();
+
+        /// <summary>
+        /// Mapping from hashing ID of a <see cref="DataViewType"/> to hashing ID of a <see cref="Type"/> and its <see cref="Attribute"/>s.
+        /// </summary>
+        private static Dictionary<int, int> _dataViewTypeIdToTypeIdMap = new Dictionary<int, int>();
+
+        private static object _lock = new object();
+
+        /// <summary>
+        /// This function computes a hashing ID from <paramref name="rawType"/> and attributes attached to it.
+        /// If a type is defined as a member in a <see langword="class"/>, <paramref name="rawTypeAttributes"/> can be obtained by calling
+        /// <see cref="MemberInfo.GetCustomAttributes(bool)"/>.
+        /// </summary>
+        /// <returns></returns>
+        private static int ComputeHashCode(Type rawType, IEnumerable<Attribute> rawTypeAttributes)
+        {
+            if (rawTypeAttributes == null)
+                return rawType.GetHashCode();
+
+            var code = rawType.GetHashCode();
+            foreach (var attr in rawTypeAttributes)
+                code = Hashing.CombineHash(code, attr.GetHashCode());
+            return code;
+        }
+
+        /// <summary>
+        /// This function hashes a <see cref="DataViewType"/> and its own hashing code together.
+        /// </summary>
+        private static int ComputeHashCode(DataViewType dataViewType) => Hashing.CombineHash(dataViewType.GetType().GetHashCode(), dataViewType.GetHashCode());
+
+        /// <summary>
+        /// Returns the <see cref="DataViewType"/> registered for <paramref name="rawType"/> and its <paramref name="rawTypeAttributes"/>.
+        /// </summary>
+        public static DataViewType GetDataViewType(Type rawType, IEnumerable<Attribute> rawTypeAttributes = null)
+        {
+            // Overall flow:
+            //   type (Type) + attrs ----> type ID ----------------> associated DataViewType's ID ----------------> DataViewType
+            //                     (hashing)      (dictionary look-up)                           (dictionary look-up)
+            lock (_lock)
+            {
+                // Compute the ID of type with extra attributes.
+                var typeId = ComputeHashCode(rawType, rawTypeAttributes);
```

Comment:
HashCodes aren't necessarily unique, we shouldn't be using them as keys in a dictionary.

Instead, we should have a "Key" type that implements GetHashCode and Equals, and we should be putting the "Key" types into the dictionary, not an integer.

Reply:
Just pushed iteration 15 for addressing this comment.

---
In reply to: [287034735](https://github.com/dotnet/machinelearning/pull/3263#discussion_r287034735) [](ancestors = 287034735)